\subsection{Determining which Components to GPU Accelerate} \label{methods:determining_which_components_to_gpu_accelerate}

When developing GKAGE, we started with KAGE as a baseline and analyzed the pipeline to find the most pronounced bottlenecks in terms of runtime.
We then examined whether these bottlenecks could benefit from GPU acceleration by taking into consideration the type of computations that were performed, whether they would be suitable for the GPU's architecture and how much of the overall runtime the component constituted. 
We would then intuit whether the component in question would be worthwhile trying to GPU accelerate given the project's time constraints and estimated difficulty.

The KAGE genotyping pipeline, which we described in section \ref{background:kage}, is split into two distinct processes (or programs): 1) counting the \textit{k}mer frequencies observed in the DNA reads of the individual being genotyped, and 2) genotyping the individual using a statistical model based on the observed and expected \textit{k}mer frequencies.
Initial benchmarking on a consumer desktop revealed that the \textit{k}mer counting step constituted nearly 97\% of KAGE's total runtime, with the \textit{k}mer counting step taking 2080 seconds and the genotyping step taking 68 seconds for a total of 2148 seconds (35 minutes and 48 seconds) to genotype a full human genome.
Thus, the \textit{k}mer counting step, having such a clear margin for runtime improvement over the genotyping step, was the main focus for GPU acceleration for potential runtime speedup.
